lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -592,27 +615,51 @@ monitorStakePools tr gp nl DBLayer{..} =
 monitorMetadata
     :: Tracer IO StakePoolLog
     -> GenesisParameters
-    -> (   PoolId
-        -> StakePoolMetadataUrl
-        -> StakePoolMetadataHash
-        -> IO (Maybe StakePoolMetadata)
-       )
     -> DBLayer IO
     -> IO ()
-monitorMetadata tr gp fetchMetadata DBLayer{..} = forever $ do
-    refs <- atomically (unfetchedPoolMetadataRefs 100)
-    successes <- fmap catMaybes $ forM refs $ \(pid, url, hash) -> do
-        fetchMetadata pid url hash >>= \case
-            Nothing -> Nothing <$ do
-                atomically $ putFetchAttempt (url, hash)
-
-            Just meta -> Just hash <$ do
-                atomically $ putPoolMetadata hash meta
-
-    when (null refs || null successes) $ do
-        traceWith tr $ MsgFetchTakeBreak blockFrequency
-        threadDelay blockFrequency
+monitorMetadata tr gp DBLayer{..} = do
+    initSettings <- atomically readSettings
+    sref <- newIORef initSettings
+    forever $ do
+        prevSettings <- readIORef sref
+        newSettings <- atomically readSettings
+        writeIORef sref newSettings
+
+        -- delete old metadata table if need be
+        case (poolMetadataSource prevSettings, poolMetadataSource newSettings) of
+            (_, FetchNone) -> liftIO $ atomically $ delPoolMetadata
+            (old, new)
+                | old /= new -> liftIO $ atomically $ delPoolMetadata
+            _ -> pure ()
+
+        -- fetch
+        (refs, successes) <- case poolMetadataSource newSettings of
+            FetchNone -> pure ([], [])
+            FetchDirect -> (fetcher Nothing) >>= fetchThem
+            FetchSMASH uri -> (fetcher $ Just uri) >>= fetchThem
```

Comment:
It's sort of weird to use a `Maybe` here that gets remapped to a fetching strategy below. Wouldn't be more transparent to simply pass the url builder strategy to the `fetcher` function ? 

```hs
FetchDirect -> fetcher identityUrlBuilder >>= fetchThem
FetchSMASH uri -> fetcher registryUrlBuilder >>= fetchThem
```

?